README.md
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3251553.svg)](https://doi.org/10.5281/zenodo.3251553)
[![Continuous Integration](https://github.com/LUMC/hutspot/actions/workflows/ci.yml/badge.svg)](https://github.com/LUMC/hutspot/actions/workflows/ci.yml)

# Hutspot

This is a multi sample DNA variant calling pipeline based on Snakemake, bwa and
the GATK HaplotypeCaller.

## Features
* Any number of samples is supported
* Whole-genome calling, regardless of wet-lab library preparation.
* Follows modern best practices
    * Each sample is individually called as as a GVCF.
    * A VCF is then produced by genotyping the individual GVCFs separately
      for each sample.
* Data parallelization for calling and genotyping steps.
    * Using the `scatter_size` setting in the configuration file, the reference
      genome is split into chunks, and each chunk can be processed
      independenly. The default value of 1 billon will scatter the human
      reference genoom into 6 chunks.
* Reasonably fast.
    * 96 exomes in < 24 hours.
* No unnecessary jobs
* Calculate coverage metrics if a `bedfile` is specified.
* Fully containerized rules through singularity and biocontainers. Legacy
conda environments are no long available.

# Installation

This repository contains a [conda](https://conda.io/docs/)
environment file that you can use to install all dependencies in a
conda environment:

```bash
conda env create -f environment.yml
```

## Singularity

We highly recommend the user of the containerized rules through
[singularity](https://www.sylabs.io/singularity/).

This option does require you to install singularity on your system. As this
usually requires administrative privileges, singularity is not contained
within our provided conda environment file.

If you want to use singularity, make sure you install version 3 or higher.

### Debian
If you happen to use Debian buster, singularity 3.0.3 comes straight out
of the box with a simple:

```bash
sudo apt install singularity-container
```

### Docker

You can run singularity within a docker container. Please note that
the container **MUST** run in privileged mode for this to work.

We have provided our own container that includes singularity and snakemake
[here](https://hub.docker.com/r/lumc/singularity-snakemake).

### Manual install

If you don't use Debian buster and cannot run a privileged docker container,
you - unfortunately :-( - will have to install singularity manually.
Please see the installation instructions
[here](https://github.com/sylabs/singularity/blob/master/INSTALL.md) on how
to do that.


## Operating system

Hutspot was tested on Ubuntu 16.04 only.
It should reasonably work on most modern Linux distributions.

# Requirements

For every sample you wish to analyze, we require one or more paired end
readgroups in fastq format. They must be compressed with either `gzip` or
`bgzip`.

The configuration must be passed to the pipeline through a configuration file.
This is a json file listing the samples and their associated readgroups
as well as the other settings to be used.
An example config json can be found [here](config/example.json), and a
json schema describing the configuration file can be found [here](config/schema.json).
This json schema can also be used to validate your configuration file.

## Reference files

The following reference files **must** be provided in the configuration:

1. `reference`: A reference genome, in fasta format. Must be indexed with
   `samtools faidx`.
2. `dbsnp`: A dbSNP VCF file
3. `known_sites`: One ore more VCF files with known sites for base
    recalibration

The following reference files **may** be provided:

1. `targetsfile`: Bed file of the targets of the capture kit. Used to calculate coverage.
2. `baitsfile`: Bed file of the baits of the capture kit. Used to calculate picard HsMetric.
3. `refflat`: A refFlat file to calculate coverage over transcripts.
4. `scatter_size`: Size of the chunks to split the variant calling into.
5. `female_threshold`: Fraction of reads between X and the autosomes to call as
    female.


# How to run

After installing and activating the main conda environment, as described above,
the pipeline can be started with:

```bash
snakemake -s Snakefile \
--use-singularity \
--configfile tests/data/config/sample_config.json
```

This would start all jobs locally. Obviously this is not what one would
regularly do for a normal pipeline run. How to submit jobs on a cluster is
described later. Let's first move on to the necessary configuration values.

## Configuration values
The required and optional outputs are specified in the json schema located in
`config/schema.json`. Before running, the content of the `--configfile` is
validated against this schema.

The following configuration values are **required**:

| configuration | description |
| ------------- | ----------- |
| `reference` | Absolute path to fasta file |
| `samples` | One or more samples, with associated fastq files |
| `dbsnp` | Path to dbSNP VCF file|
| `known_sites` | Path to one or more VCF files with known sites. Can be the same as the `dbsnp` file|


The following configuration options are **optional**:

| configuration | description |
| ------------- | ----------- |
| `targetsfile` | Bed file of the targets of the capture kit. Used to calculate coverage |
| `baitsfile` | Bed file of the baits of the capture kit. Used to calculate picard HsMetrics |
| `female_threshold` | Float between 0 and 1 that signifies the threshold of the ratio between coverage on X/overall coverage that 'calls' a sample as female. Default = 0.6 |
| `scatter_size` | The size of chunks to divide the reference into for parallel execution. Default = 1000000000 |
| `coverage_threshold` | One or more threshold coverage values. For each value, a sample specific bed file will be created that contains the regions where the coverage is above the threshold |
| `restrict_BQSR` | Restrict GATK BaseRecalibration to a single chromosome. This is faster, but the recalibration is possibly less reliable |
| `multisample_vcf` | Create a true multisample VCF file, in addition to the regular per-sample VCF files |


## Cluster configuration

To run on a cluster, snakemake needs to be called with some extra arguments.
Additionally, it needs a cluster yaml file describing resources per job.

If you run on a cluster with drmaa support,an environment variable named
`DRMAA_LIBRARY_PATH` must be in the executing shell environment. This variable
points to the `.so` file of the DRMAA library.

An sge-cluster.yml is bundled with this pipeline in the cluster directory.
It is optimized for SGE clusters, where the default vmem limit is 4G.
If you run SLURM, or any other cluster system, you will have to write your own
cluster yaml file. Please see the [snakemake documentation](http://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html#cluster-configuration)
for details on how to do so. Given the provided sge-cluster.yml, activating the
cluster mode can be done as follows:

```bash
snakemake -s Snakefile \
--cluster-config cluster/sge-cluster.yml
--drmaa ' -pe <PE_NAME> {cluster.threads} -q all.q -l h_vmem={cluster.vmem} -cwd -V -N hutspot' \
```

## Limitations
Sample names should be unique, and not overlap (such as `sample1` and
`sample10`). This is due to the way output files are parsed by multiQC,
when sample names overlap, the json output for picard DuplicationMetrics cannot
be parsed unambiguously.

## Binding additional directories under singularity

In singularity mode, snakemake binds the location of itself in the container.
The current working directory is also visible directly in the container.

In many cases, this is not enough, and will result in `FileNotFoundError`s.
E.g., suppose you run your pipeline in `/runs`, but your fastq files live in
`/fastq` and your reference genome lives in `/genomes`. We would have to bind
`/fastq` and `/genomes` in the container.

This can be accomplished with `--singularity-args`, which accepts a simple
string of arguments passed to singularity. E.g. in the above example,
we could do:

```bash
snakemake -S Snakefile \
--use-singularity  \
--singularity-args ' --bind /fastq:/fastq --bind /genomes:/genomes '
```

## Summing up

To sum up, a full pipeline run under a cluster would be called as:

```bash
snakemake -s Snakefile \
--use-singularity \
--singularity-args ' --bind /some_path:/some_path ' \
--cluster-config cluster/sge-cluster.yml \
--drmaa ' -pe <PE_NAME> {cluster.threads} -q all.q -l h_vmem={cluster.vmem} -cwd -V -N hutspot' \
--rerun-incomplete \
--jobs 200 \
-w 120 \
--max-jobs-per-second 30 \
--restart-times 2 \
--configfile config.json
```

# Graph

Below you can see the rule graph of the pipeline. The main variant calling flow
is highlighted in red. This only shows dependencies
between rules, and not between jobs. The actual job graph is considerably
more complex, as nearly all rules are duplicated by sample and some
(the scatter jobs) additionally by chunk.

As a rough estimate of the total number of jobs in pipeline you can use
the following formula:

<a href="https://www.codecogs.com/eqnedit.php?latex=n_{jobs}&space;=&space;4&plus;(21*n_{samples})&plus;(n_{samples}*n_{beds})&plus;(n_{samples}*n_{chunks})&plus;n_{chunks}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?n_{jobs}&space;=&space;4&plus;(21*n_{samples})&plus;(n_{samples}*n_{beds})&plus;(n_{samples}*n_{chunks})&plus;n_{chunks}" title="n_{jobs} = 4+(21*n_{samples})+(n_{samples}*n_{beds})+(n_{samples}*n_{chunks})+n_{chunks}" /></a>

<!---
Note: math doesn't work on github. The following _does_ work in gitlab
```math
jobs = 4+(21*n_{samples})+(1*n_{samples}*n_{beds})+(1*n_{samples}*n_{chunks})+(1*n_{chunks})
```
--->

This gives about 12,000 jobs for a 96-sample run with 2 bed files and 100 chunks.

NOTE: the graph will only render if your markdown viewer supports `plantuml`.
Having trouble viewing the graph? See [this](img/rulegraph.svg) static SVG instead.

```plantuml
digraph snakemake_dag {
    graph[bgcolor=white, margin=0];
    node[shape=box, style=rounded, fontname=sans,                 fontsize=10, penwidth=2];
    edge[penwidth=2, color=grey];
	0[label = "all", color = "0.30 0.6 0.85", style="rounded"];
	1[label = "multiqc", color = "0.60 0.6 0.85", style="rounded"];
	2[label = "merge_stats", color = "0.17 0.6 0.85", style="rounded"];
	3[label = "bai", color = "0.09 0.6 0.85", style="rounded"];
	4[label = "genotype_gather\nsample: micro", color = "0.06 0.6 0.85", style="rounded"];
	5[label = "gvcf_gather\nsample: micro", color = "0.32 0.6 0.85", style="rounded"];
	6[label = "fastqc_raw\nsample: micro", color = "0.00 0.6 0.85", style="rounded"];
	7[label = "fastqc_merged", color = "0.11 0.6 0.85", style="rounded"];
	8[label = "fastqc_postqc", color = "0.02 0.6 0.85", style="rounded"];
	9[label = "stats_tsv", color = "0.45 0.6 0.85", style="rounded"];
	10[label = "collectstats", color = "0.24 0.6 0.85", style="rounded"];
	11[label = "vcfstats\nsampel: micro", color = "0.52 0.6 0.85", style="rounded"];
	12[label = "markdup", color = "0.47 0.6 0.85", style="rounded"];
	13[label = "scatterregions", color = "0.56 0.6 0.85", style="rounded"];
	14[label = "merge_r1\nsample: micro", color = "0.65 0.6 0.85", style="rounded"];
	15[label = "merge_r2\nsample: micro", color = "0.26 0.6 0.85", style="rounded"];
	16[label = "cutadapt", color = "0.22 0.6 0.85", style="rounded"];
	17[label = "fqcount_preqc", color = "0.37 0.6 0.85", style="rounded"];
	18[label = "fqcount_postqc", color = "0.58 0.6 0.85", style="rounded"];
	19[label = "mapped_reads_bases", color = "0.43 0.6 0.85", style="rounded"];
	20[label = "unique_reads_bases", color = "0.34 0.6 0.85", style="rounded"];
	21[label = "fastqc_stats", color = "0.13 0.6 0.85", style="rounded"];
	22[label = "covstats", color = "0.39 0.6 0.85", style="rounded"];
	23[label = "align", color = "0.49 0.6 0.85", style="rounded"];
	24[label = "create_markdup_tmp", color = "0.41 0.6 0.85", style="rounded,dashed"];
	25[label = "sickle", color = "0.19 0.6 0.85", style="rounded"];
	26[label = "genome", color = "0.62 0.6 0.85", style="rounded"];
	1 -> 0
	2 -> 0
	3 -> 0
	4 -> 0
	5 -> 0
	6 -> 0
	7 -> 0
	8 -> 0
	9 -> 1
	10 -> 2
	11 -> 2
	12 -> 3
	13 -> 4
	13 -> 5
	14 -> 7
	15 -> 7
	16 -> 8
	2 -> 9
	17 -> 10
	18 -> 10
	19 -> 10
	20 -> 10
	21 -> 10
	22 -> 10
	4 -> 11
	23 -> 12
	24 -> 12
	25 -> 16
	14 -> 17
	15 -> 17
	16 -> 18
	23 -> 19
	12 -> 20
	7 -> 21
	8 -> 21
	12 -> 22
	26 -> 22
	16 -> 23
	24 -> 23
	14 -> 25
	15 -> 25
}
```

LICENSE
=======

AGPL-3.0


CHANGELOG.md
Changelog
==========

<!--
Newest changes should be on top.

This document is user facing. Please word the changes in such a way
that users understand how the changes affect the new version.
-->

v2.2.2-dev
---------------------------

v2.0.1
---------------------------
+ `multisample_vcf` now acts on the scatters, instead of on the merged g.vcf
files.
+ The multisample output is located in `multisample/multisample.vcf.gz`.
+ Intermediate .bam, .bai and fastq files are automatically removed when no
longer needed.
+ Switch to using chunked-scatter

v2.0.0
---------------------------
+ Add an environment.yml file for conda.
+ Greatly simplified the snakemake workflow.
+ All statistics are now calculated using existing tools.
+ Add option `multisample_vcf` to enable joint variantcalling.


src/covstats.py
#!/usr/bin/env python2.7

"""
Plot output of coverageBed / "bedtools coverage" when run with the -d flag.


This script plots a bar graph showing how many times a base covered X times
are found. An additional box plot is also plotted to show the general trend
of the coverage.

By default, the bar graph only shows the data up to the 98th percentile.
Counts for bases that are not covered (0x coverage) are shown in lighter
shade. The median, quartile, and whiskers values shown in the boxplot are
still computed from the complete data set (0x coverage included).

Requirements:
    * Python == 2.7.x
    * Matplotlib >= 1.3.0
    * Numpy >= 1.8.0

Copyright (c) 2013 Wibowo Arindrarto <w.arindrarto@lumc.nl>
Copyright (c) 2013 LUMC Sequencing Analysis Support Core <sasc@lumc.nl>
MIT License <http://opensource.org/licenses/MIT>
"""

RELEASE = False
__version_info__ = ('0', '1',)
__version__ = '.'.join(__version_info__)
__version__ += '-dev' if not RELEASE else ''

import argparse
import collections
import itertools
import json
import locale
import os
import sys

import matplotlib

matplotlib.use('Agg')
import matplotlib.gridspec as gs
import matplotlib.pyplot as plt
import matplotlib.ticker as tkr
import numpy as np

BLUE = '#2166AC'
RED = '#1A9850'
GREEN = '#D6604D'

locale.setlocale(locale.LC_ALL, '')
group_digits = lambda x, pos: locale.format('%d', x, grouping=True)
major_formatter = tkr.FuncFormatter(group_digits)


def cachedproperty(func):
    """Decorator for cached property loading."""
    attr_name = func.__name__

    @property
    def cached(self):
        if not hasattr(self, '_cache'):
            setattr(self, '_cache', {})
        try:
            return self._cache[attr_name]
        except KeyError:
            result = self._cache[attr_name] = func(self)
            return result

    return cached


class Coverage(object):
    """Class representing coverage metrics from a coverageBed -d output."""

    def __init__(self, cvg_list, name='?'):
        """

        :param cvg_list: iterator yielding coverage per position
        :type cvg_list: iterable

        """
        assert cvg_list
        counter = collections.Counter()
        total_bases, nonzero_bases = 0, 0
        for cvg in cvg_list:
            counter[cvg] += 1
            total_bases += 1
            if cvg > 0:
                nonzero_bases += 1

        self._counter = counter
        self.total_bases = total_bases
        self.nonzero_bases = nonzero_bases

    def __iter__(self):
        return self._counter.iteritems()

    def __repr__(self):
        return "{0}(...)".format(self.__class__.__name__)

    @cachedproperty
    def total(self):
        """Total coverage."""
        return sum([cvg * count for cvg, count in self])

    @cachedproperty
    def horizontal(self):
        """Horizontal coverage (0 - 1)."""
        return float(self.nonzero_bases) / self.total_bases

    @cachedproperty
    def mean(self):
        """Average coverage."""
        return float(self.total) / self.total_bases

    @cachedproperty
    def max(self):
        """Maximum coverage."""
        return max(self._counter.keys())

    @cachedproperty
    def median(self):
        """Median coverage."""
        return np.percentile(self.cov_counts, 50)

    @cachedproperty
    def mode(self):
        """Modal coverage."""
        count = collections.Counter(self.cov_counts)
        # return self.cov_counts[self.cov_counts.index(self.max)]
        return count.most_common(1)[0][0]

    @cachedproperty
    def cov_counts(self):
        """List of coverage for each base position (unsorted)."""
        # get list of each base coverage
        cov_counts = ([key] * value for key, value in self._counter.items())
        # and flatten the list
        return [item for sublist in cov_counts for item in sublist]

    def at_least(self, n):
        """Return the percentages of bases covered at least n times."""
        x = sum([count for cvg, count in self if cvg >= n])
        return float(x) / self.total_bases

    def get_quick_stats(self):
        """Returns a dictionary containing quick coverage statistics."""
        return {
            'max': self.max,
            'median': self.median,
            'mode': self.mode,
            'mean': self.mean,
            'horizontal': self.horizontal,
            'width': self.total_bases,
            'width_nonzero': self.nonzero_bases,
            'total': self.total,
            'frac_min_10x': self.at_least(10),
            'frac_min_20x': self.at_least(20),
            'frac_min_30x': self.at_least(30),
            'frac_min_40x': self.at_least(40),
            'frac_min_50x': self.at_least(50),
            'frac_min_100x': self.at_least(100),
        }

    def plot(self, min_cov_ok=7, percentile_show=98, title=None, out_img=None):
        """Plots the coverage object.

        :param min_cov_ok: Minimum coverage value to show in the bar graph
            (default: 7).
        :type min_cov_ok: int
        :param percentile_show: Maximum percentile to show in the bar graph
            (default: 98).
        :type percentile_show: int
        :param title: Plot title
        :type title: list of strings (one item per line)
        :param out_img: Output image filename.
        :type out_img: str

        """
        plt.figure(figsize=(8, 8))
        grids = gs.GridSpec(2, 1, height_ratios=[5, 1])

        ax0 = plt.subplot(grids[0])

        if title is None:
            title = ['Coverage Plot']
        elif isinstance(title, basestring):
            title = [title]
        t = plt.title('\n'.join(title), fontsize=20)
        t.set_y(1.05)

        x_data, y_data = [], []
        x_data_shade, y_data_shade = [], []
        for coverage, count in self:
            if coverage >= min_cov_ok:
                x_data.append(coverage)
                y_data.append(count)
            else:
                x_data_shade.append(coverage)
                y_data_shade.append(count)

        ax0.bar(x_data, y_data, width=1, linewidth=0, color=BLUE, align='center')
        ax0.yaxis.set_major_formatter(major_formatter)
        ax0.grid(True)
        plt.ylabel('Counts')
        ax0.text(0.5, 0.95, 'Mean: %.2fx     Median: %.2fx     '
                            'Horizontal: %.2f%%' % (self.mean, self.median, 100.0 * \
                                                    self.horizontal), verticalalignment='center',
                 horizontalalignment='center', transform=ax0.transAxes,
                 bbox=dict(facecolor='wheat', alpha=0.5, boxstyle='round'),
                 size=12)

        ax1 = plt.subplot(grids[1], sharex=ax0)
        ax1.axes.get_yaxis().set_visible(False)
        bp = ax1.boxplot(self.cov_counts, vert=False, widths=0.6, sym='+')
        for x in itertools.chain(bp['boxes'], bp['medians'], bp['whiskers'],
                                 bp['caps']):
            x.set(color=BLUE, linewidth=1.6)

        for flier in bp['fliers']:
            flier.set(color=RED, alpha=0.5)
        # bp['fliers'][0].set(color=RED, alpha=0.5)
        # bp['fliers'][1].set(color=GREEN, alpha=0.5)

        upper_limit = np.percentile(self.cov_counts, percentile_show)
        if x_data:
            space = (upper_limit - min(x_data)) / 40
        else:
            space = 0
        # truncate plot if we're not displaying maximum value
        if upper_limit != self.max:
            ax1.set_xlim([min(self.cov_counts) - space - 0.5, upper_limit + 0.5])
        # otherwise, give some space
        else:
            ax1.set_xlim([min(self.cov_counts) - space - 0.5, upper_limit +
                          space + 0.5])

        # plot shaded values
        if x_data_shade:
            ylim = ax0.get_ylim()
            ax0.bar(x_data_shade, y_data_shade, width=1, linewidth=0,
                    color=BLUE, alpha=0.3, align='center')
            ax0.set_ylim(ylim)

        plt.xlabel('Coverage')
        grids.update(hspace=0.075)

        if out_img is not None:
            plt.savefig(out_img, bbox_inches='tight')


if __name__ == '__main__':

    usage = __doc__.split('\n\n\n')
    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description=usage[0], epilog=usage[1])

    parser.add_argument('input', type=str, help='Path to input file '
                                                '(coverageBed output) or \'-\' for stdin')
    parser.add_argument('--plot', dest='plot', type=str,
                        help='Path to output PNG file')
    parser.add_argument('--min-cov-show', dest='min_cov_ok', type=int,
                        default=6, help='Minimum coverage to show in bar graph')
    parser.add_argument('--max-percentile-show', dest='max_pct_show', type=int,
                        default=98, help='Maximum percentile to show in bar graph')
    parser.add_argument('--title', dest='title', type=str,
                        default='Coverage Plot', help='Plot title')
    parser.add_argument('--subtitle', dest='subtitle', type=str, help='Plot subtitle')

    args = parser.parse_args()

    if args.input == '-':
        instream = sys.stdin
    else:
        instream = open(args.input, 'r')

    title = [args.title]
    if args.subtitle is None:
        title.append("'" + args.input + "'")
    else:
        title.append(args.subtitle)

    coverages = {}
    all_covs, cur_covs = [], []
    cur_chrom, prev_chrom = None, None
    for line in (l.strip() for l in instream):
        cols = line.split('\t')
        cur_chrom = cols[0]
        cvg = int(cols[-1])
        # coverage for all positions
        all_covs.append(cvg)
        # coverage per chromosome
        if cur_chrom != prev_chrom:
            if prev_chrom is not None:
                coverages[prev_chrom] = cur_covs
            cur_covs = []
            prev_chrom = cur_chrom
        cur_covs.append(cvg)
    # also append the last chromosome from the file
    coverages[cur_chrom] = cur_covs

    coverages['_all'] = all_covs

    for cname, clist in coverages.items():
        coverages[cname] = Coverage(clist)

    if args.input != '-':
        instream.close()

    if args.plot is not None:
        coverages['_all'].plot(min_cov_ok=args.min_cov_ok, percentile_show=args.max_pct_show,
                               title=title, out_img=args.plot)

    stats = {'coverage': {k: v.get_quick_stats() for k, v in coverages.items()}}
    if args.plot is not None:
        files = {'plot_coverage': {
            'path': os.path.abspath(args.plot),
            'checksum_sha1': None,
        }
        }
    else:
        files = {}

    json.dump({'stats': stats, 'files': files}, sys.stdout, sort_keys=True, indent=4,
              separators=(',', ': '))

